### PR TITLE
Chainweb.Version recordification and overhaul

### PR DESCRIPTION
**Pull request by edmundnoble**
Integration test replay is expected to fail, because devnet parameters change. I will follow up by running the integration-tests before merge. EDIT: that is no longer true, the devnet parameters are now the same and devnet integration tests replay fine.

### CODE REVIEW

File: bench/Chainweb/Pact/Backend/Bench.hs [hunk 1]
```diff
@@ -45,7 +45,7 @@ import Chainweb.Utils.Bench
 import Chainweb.Utils (sshow)
 import Chainweb.Version
 
-v :: ChainwebVersion
+v :: ChainwebVersion dc
```
Comment (jwiegley) [inline bench/Chainweb/Pact/Backend/Bench.hs:48]: What do the letters `dc` mean?

File: test/Chainweb/Test/BlockHeader/Validation.hs [hunk 2]
```diff
@@ -139,21 +139,21 @@ prop_validateHeader msg h = testCase msg $ do
...
-prop_fail_validate = validate_cases "validate invalid BlockHeaders" validationFailures
+prop_fail_validate = validate_cases Mainnet01 "validate invalid BlockHeaders" validationFailures
 
 prop_da_validate :: TestTree
-prop_da_validate = validate_cases "difficulty adjustment validation" daValidation
+prop_da_validate = validate_cases (Development defaultDevVersionConfig) "difficulty adjustment validation" daValidation
 
 prop_legacy_da_validate :: TestTree
-prop_legacy_da_validate = validate_cases "legacy difficulty adjustment validation" legacyDaValidation
+prop_legacy_da_validate = validate_cases Mainnet01 "legacy difficulty adjustment validation" legacyDaValidation
```
Comment (jwiegley) [inline test/Chainweb/Test/BlockHeader/Validation.hs:148]: Why not `Development` here?

File: test/Chainweb/Test/Mempool/Consensus.hs [hunk 3]
```diff
@@ -354,7 +356,7 @@ header' h = do
             :+: MerkleLogBody mempty
    where
     BlockCreationTime t = _blockCreationTime h
-    target = powTarget (ParentHeader h) mempty t'
+    target = powTarget (v & versionConfig .~ undefined) (ParentHeader h) mempty t'
```
Comment (jwiegley) [inline test/Chainweb/Test/Mempool/Consensus.hs:359]: Um, what is the purpose of this?

Comment (jwiegley) [reply test/Chainweb/Test/Mempool/Consensus.hs:359]: Like you do elsewhere, I think calling `error` here is much to be preferred.

File: bench/Chainweb/Pact/Backend/Bench.hs (hunk 1 above)
Comment (edmundnoble) [reply bench/Chainweb/Pact/Backend/Bench.hs:48]: Devnet config.

Comment (edmundnoble) [reply bench/Chainweb/Pact/Backend/Bench.hs:49]: I don't really like this name, how bout `c`? or `cfg`?

File: src/Chainweb/BlockHeader.hs [hunk 4]
```diff
@@ -468,150 +617,140 @@ instance HasChainGraph ParentHeader where
...
+        [ InputNode "CHAINWEB_GENESIS"
+        , encodeMerkleInputNode encodeChainwebVersionCode (_versionCode v)
+        , encodeMerkleInputNode encodeChainId (_chainId p)
+        ]
 
-instance Ord BlockHeader where
-     compare = compare `on` _blockHash
-     {-# INLINE compare #-}
+{-# NOINLINE genesisBlockHeaderCache #-}
+genesisBlockHeaderCache :: IORef (HashMap ChainwebVersionCode (HashMap ChainId BlockHeader))
```
Comment (larskuhtz) [inline src/Chainweb/BlockHeader.hs:638]: We may consider doing a more efficient memoization for production versions (in particular with large histories, where we may do long replays or catchups).

File: src/Chainweb/BlockHeader.hs [hunk 5]
```diff
@@ -468,150 +617,140 @@ instance HasChainGraph ParentHeader where
...
+-- | A block chain is globally uniquely identified by its genesis hash.
+-- Internally, we use the 'ChainwebVersionTag value and the 'ChainId'
+-- as identifiers. We thus include the 'ChainwebVersionTag value and the
+-- 'ChainId' into the genesis block hash.
+--
+-- We assume that there is always only a single 'ChainwebVersionTag in
+-- scope and identify chains only by their internal 'ChainId'.
+--
+genesisBlockHeaders :: ChainwebVersion -> HashMap ChainId BlockHeader
+genesisBlockHeaders v = unsafePerformIO $
```
Comment (larskuhtz) [inline src/Chainweb/BlockHeader.hs:655]: consider using `unsafeDupablePerformIO` instead `unsafePerformIO`.

File: src/Chainweb/BlockHeader.hs [hunk 6]
```diff
@@ -211,19 +233,172 @@ encodeEpochStartTime (EpochStartTime t) = encodeTime t
...
+            -- of a block is the block weight of the parent plus the difficulty
+            -- of the block.
+
+        , _blockHeight :: {-# UNPACK #-} !BlockHeight
+            -- ^ block height records the length of the chain. It is redundant
+            -- information and thus subject the inductive property that
+            -- the block height of a block is the block height of its parent
+            -- plus one.
+
+        , _blockChainwebVersion :: !ChainwebVersion
```
Comment (larskuhtz) [inline src/Chainweb/BlockHeader.hs:342]: Consider `ChainwebVersionCode` and define an `HasChainwebVersion` instance for `ChainwebVersionCode`.

File: src/Chainweb/Chainweb/Configuration.hs [hunk 7]
```diff
@@ -555,4 +559,40 @@ pChainwebConfiguration = id
         % long "module-cache-limit"
         <> help "Maximum size of the per-chain checkpointer module cache in bytes"
         <> metavar "INT"
+    where
+    knownVersion = do
+        option (findKnownVersion =<< textReader)
+            % long "chainweb-version"
+            <> short 'v'
+            <> help "the chainweb version that this node is using"
+    version = constructVersion
```
Comment (larskuhtz) [inline src/Chainweb/Chainweb/Configuration.hs:568]: Consider using the `ConfigValidation` and define the command line parse for Version as separate function.

File: test/Chainweb/Test/TestVersions.hs [hunk 8]
```diff
@@ -0,0 +1,280 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE RecursiveDo #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Chainweb.Test.TestVersions
```
Comment (larskuhtz) [inline test/Chainweb/Test/TestVersions.hs:9]: Please document how test versions are constructed. Also annotate each constructor field that must be lazy.

File: test/Chainweb/Test/TestVersions.hs [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+            { _disableMempoolSync = True
+            , _disablePeerValidation = True
+            }
+        & versionGenesis .~ VersionGenesis
+            { _genesisBlockPayload = AllChains emptyPayload
+            , _genesisBlockTarget = AllChains maxTarget
+            , _genesisTime = AllChains $ BlockCreationTime epoch
+            }
+        & versionForks .~ fastForks
+        & versionUpgrades .~ forkUpgrades v
```
Comment (larskuhtz) [inline test/Chainweb/Test/TestVersions.hs:159]: Leave that empty

**Comment by edmundnoble** [comment]
Adding Stuart here for the Pact service stuff, specifically in TransactionExec.

File: bench/Chainweb/Pact/Backend/ForkingBench.hs [hunk 10]
```diff
@@ -379,7 +379,7 @@ cid :: ChainId
 cid = someChainId testVer
 
 testVer :: ChainwebVersion
-testVer = FastTimedCPM petersonChainGraph
+testVer = slowForkingCpmTestVersion petersonChainGraph
```
Comment (chessai) [inline bench/Chainweb/Pact/Backend/ForkingBench.hs:382]: Can you explain this change to me?

Comment (edmundnoble) [reply bench/Chainweb/Pact/Backend/ForkingBench.hs:382]: Historically we special-cased the fork heights of `FastTimedCPM petersonChainGraph` to be slower, so any other chain graph gives faster fork heights. Confusing, eh? With this PR, you ask for the slower or faster fork heights, and you ask for the chain graph separately.

tl;dr:
`FastTimedCPM petersonChainGraph` corresponds to `slowForkingCpmTestVersion petersonChainGraph`, whereas `FastTimedCPM g` for `g /= petersonChainGraph` corresponds to `fastForkingCpmTestVersion petersonChainGraph`.

Comment (edmundnoble) [reply bench/Chainweb/Pact/Backend/ForkingBench.hs:382]: For example:
```hs
pact420Upgrade :: ChainwebVersion -> BlockHeight -> Bool
pact420Upgrade Mainnet01 = (>= 2_334_500) -- 2022-01-17T17:51:12
pact420Upgrade Testnet04 = (>= 1_862_000) -- 2022-01-13T16:11:10
pact420Upgrade Development = (>= 90)
pact420Upgrade (FastTimedCPM g) | g == petersonChainGraph = (>= 5)
pact420Upgrade _ = const True
```

**Comment by larskuhtz** [comment]
@edmundnoble The current version of this PR fails devnet replay validation with `Description: The POW hash does not match the POW target of the block`

**Comment by edmundnoble** [comment]
@larskuhtz fixed, replay going on at https://github.com/kadena-io/integration-tests/actions/runs/5316476997/jobs/9626071632.

**Comment by edmundnoble** [comment]
Replay for [4acf057](https://github.com/kadena-io/chainweb-node/pull/1550/commits/4acf057d3388b71d1d1f50f6689d6a6da26f6c4e) passed.

**Review by larskuhtz — APPROVED** [review]

**Review by chessai — APPROVED** [review]
